### PR TITLE
Update k8s scheduler lib and report packing metrics

### PR DESCRIPTION
**Pull request by pisarenko-net**
PR implements new metrics for packing efficiency. Packing is now run with 2 functions: the configured one and "pack tightly". The second round of packing is only used for reporting in metrics.

This depends on a new (unreleased) version of k8s-spark-scheduler-lib so this can't be merged until a new lib is released and dependencies are updated.

### CODE REVIEW

**Comment by pisarenko-net** [comment]
Disclaimer: please review the metrics interactions carefully. I am not sure if anything else is required (like an explicit emit, or rate limiting). If there's a good primer on metrics please send me a link.

File: internal/extender/resource.go [hunk 1]
```diff
@@ -304,14 +306,27 @@ func (s *SparkSchedulerExtender) selectDriverNode(ctx context.Context, driver *v
...
+		ctx,
+		applicationResources.driverResources,
+		applicationResources.executorResources,
+		applicationResources.minExecutorCount,
+		driverNodeNames,
+		executorNodeNames,
+		availableNodesSchedulingMetadata)
+	driverNode, executorNodes, hasCapacity := packingResultDefault.DriverNode, packingResultDefault.ExecutorNodes, packingResultDefault.HasCapacity
+	// as an experiment, binpack with tight pack (report in metrics but don't use this result)
+	packingResultTight := SelectBinpacker("tightlyPack").BinpackFunc(
```
Comment (Chrisbattarbee) [inline internal/extender/resource.go:321]: I think we can do this but lets keep the PR tightly scoped and do this in another PR

File: internal/metrics/binpack.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+const (
+	packingEfficiencyMetricName = "foundry.spark.scheduler.packing_efficiency"
+
+	packingResourceTagKey = "foundry.spark.scheduler.packing_resource"
+	cpuTagValue           = "CPU"
+	memoryTagValue        = "Memory"
+	gpuTagValue           = "GPU"
+	maxTagValue           = "Max"
+
+	packingEfficiencyFunctionNameTagKey = "foundry.spark.scheduler.packing_function"
```
Comment (Chrisbattarbee) [inline internal/metrics/binpack.go:57]: To be consistent we should not have `_`

File: internal/metrics/binpack.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+func reportableEfficiencyFromAvgPackingEfficiency(efficiency *binpack.AvgPackingEfficiency) reportableEfficiency {
+	return reportableEfficiency{
+		CPU:    efficiency.CPU,
+		Memory: efficiency.Memory,
+		GPU:    efficiency.GPU,
+	}
+}
+
+const (
+	packingEfficiencyMetricName = "foundry.spark.scheduler.packing_efficiency"
```
Comment (Chrisbattarbee) [inline internal/metrics/binpack.go:49]: To be consistent we should not have `_`

File: internal/metrics/binpack.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+)
+
+// ReportPackingEfficiency report packing efficiency metrics for a single packing result.
+func ReportPackingEfficiency(
+	ctx context.Context,
+	packingFunctionName string,
+	nodesSchedulingMetadata resources.NodeGroupSchedulingMetadata,
+	packingResult *binpack.PackingResult) {
+
+	contextTags := metrics.TagsFromContext(ctx)
```
Comment (Chrisbattarbee) [inline internal/metrics/binpack.go:78]: I think its unlikely that we'll want all of these tags, the cardinality of the metric will be massive

File: internal/metrics/binpack.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	registry := metrics.NewRootMetricsRegistry()
```
Comment (Chrisbattarbee) [inline internal/metrics/binpack.go:79]: We dont want to create a metrics registry here

File: internal/metrics/binpack.go [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+	packingFunctionTag := metrics.MustNewTag(packingEfficiencyFunctionNameTagKey, packingFunctionName)
+
+	// report packing efficiencies per node
+	for _, efficiency := range packingResult.PackingEfficiencies {
+		hostTag := HostTag(ctx, efficiency.NodeName)
```
Comment (Chrisbattarbee) [inline internal/metrics/binpack.go:85]: Tagging by host is going to have extremely high cardinality

File: internal/metrics/binpack.go [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+		emitMetrics(registry, contextTags, hostTag, packingFunctionTag, reportableEfficiencyFromPackingEfficiency(efficiency))
+	}
+
+	// report avg packing efficiency for all nodes at once
+	efficiency := computeAvgPackingEfficiencyForResult(nodesSchedulingMetadata, packingResult)
+	avgHostTag := HostTag(ctx, avgEfficiencyNodeName)
+	emitMetrics(registry, contextTags, avgHostTag, packingFunctionTag, reportableEfficiencyFromAvgPackingEfficiency(&efficiency))
```
Comment (Chrisbattarbee) [inline internal/metrics/binpack.go:92]: Im a bit confused as to why we're reporting twice here?

**Review by Chrisbattarbee — CHANGES_REQUESTED** [review]

File: internal/metrics/binpack.go (hunk 7 above)
Comment (pisarenko-net) [reply internal/metrics/binpack.go:92]: This is the pattern we've used in scaler telemetry: report efficiency per node and then separately for all nodes at once. The upside of reporting by node is that we can then see what the efficiency in every node. Obviously, comes with an increase in amount of data. We could start with just the avg efficiency, disregarding node stats.

**Comment by pisarenko-net** [comment]
Thank you for the review Chris, I'm going to fix merge conflicts and incorporate feedback.

File: internal/extender/resource.go (hunk 1 above)
Comment (pisarenko-net) [reply internal/extender/resource.go]: Ack!

File: internal/metrics/binpack.go (hunk 4 above)
Comment (pisarenko-net) [reply internal/metrics/binpack.go:78]: Do I understand correctly that the important tags, such as instance-group, deployment are populated automatically when calling `metrics.FromContext(ctx)...`?

**Review by Chrisbattarbee — APPROVED** [review]
